pkg/kubelet/prober/prober.go

```diff
@@ -75,11 +73,20 @@ func newProber(
 		startupHTTP:   httpprobe.New(followNonLocalRedirects),
 		tcp:           tcpprobe.New(),
 		runner:        runner,
-		refManager:    refManager,
 		recorder:      recorder,
 	}
 }
 
+// recordContainerEvent should be used by the prober for all container related events.
```

Comment:
Precisely! I just trimmed it down slightly because there was no need for the container ID vs name substitution since we don't have it in the input to the function.

In fact, wrt to your earlier comment, it's when I discovered this function existed that I figured it seemed safe to remove RefManager. I did try to hunt down some of the history, however the earliest commit of container_reference_manager.go appears to be moving the code from elsewhere according to the log comment. I then didn't try to go down the rabbit hole and find where it came from nor when it was introduced.

I was concerned that it might be some kind of optimisation, however seeing kuberuntime_container.go doesn't bother with it, I thought maybe prober.go could too.

Reply:
Travelling back in time (newest to oldest commits):

This commit moved the RefManager to pkg/kubelet/container so it could be used by other packages (says the log):
https://github.com/kubernetes/kubernetes/commit/9f62b929b885

This commit turned the previous ref manager map into the RefManager:
https://github.com/kubernetes/kubernetes/commit/a06bc964143
"This enables pluggable container runtime to manage the container references."

This commit changes the dockerIDToRef map into generic containerIDToRef map:
https://github.com/kubernetes/kubernetes/commit/027f6854d0f

This commit introduces the dockerIDToRef map:
https://github.com/kubernetes/kubernetes/commit/dcc111bf13d
Comment is: "Start putting event generation into kubelet (not enabled yet)"

The reason for creating this reference manager appears to be "Needed to report events for containers belonging to deleted/modified pods." (https://github.com/kubernetes/kubernetes/commit/dcc111bf13d#diff-bf28da68f62a8df6e99e447c4351122dR116) That comment has since disappeared in the 5 year interim.

Also, the refs are cleared from the map as soon as the containers are deleted, and nowhere other than in the prober was GetRef used other than for checking if the ref existed prior to deleting it.

So it was introduced e.g. for performance reasons.

It looks unnecessary to me, though I do agree a smaller fix can be achieved by defaulting to using the RefManager and creating the ref if it isn't there.

What do you prefer I do?

Reply:
Given that I don't have the context on why `RefManager` was first introduced, I think I'd prefer leaving `RefManager` as is, and just creating the ref if it isn't there.

That said I don't feel super strongly.

Curious to hear from other community members and @vishh.

Reply:
As we've not had any feedback, I'll split this into 2 PRs: one that just fixes the bug by using the ref if it is available (this PR), and another that proposes to remove RefManager.

Reply:
Hi @mattjmcnaughton !

I split the PR in 2 and this one now checks if the ref exists before falling back to regenerating it. What do you make of this version?

Reply:
That sounds like a good plan!